output/http/outputhttp.go

```diff
@@ -98,29 +133,121 @@ func (t *OutputConfig) Output(ctx context.Context, event logevent.LogEvent) (err
+		return ErrPermanentError.New(nil, url, resp.StatusCode)
+	}
+	if !t.checkIntInAcceptedList(resp.StatusCode) {
+		t.failedDelivery(ctx, event)
+		return ErrSoftError.New(nil, url, resp.StatusCode)
+	}
+	// the event was sent correctly, we now have to resume inputs if we have requested a pause.
+	if atomic.CompareAndSwapUint32(&t.isInPause, statusPaused, statusDelivering) {
+		goglog.Logger.Debug("outputhttp requesting resume")
+		err := t.control.RequestResume(ctx)
+		if err != nil {
+			goglog.Logger.Error("outputhttp", err.Error())
+		}
 	}
-
 	return nil
 }
 
-// checkIntInList checks if code is in configured list of accepted status codes
-func (t *OutputConfig) checkIntInList(code int) bool {
+// checkIntInAcceptedList checks if code is in configured list of accepted status codes
+func (t *OutputConfig) checkIntInAcceptedList(code int) bool {
 	for _, v := range t.AcceptedHttpResult {
 		if v == code {
 			return true
 		}
 	}
 	return false
 }
+
+// checkIntInAcceptedList checks if code is in configured list of status codes where we should discard the message
+func (t *OutputConfig) checkIntInDiscardedList(code int) bool {
+	for _, v := range t.PermanentHttpErrors {
+		if v == code {
+			return true
+		}
+	}
+	return false
+}
+
+// failedDelivery receives an event that failed delivery and triggers a pause event if we have not done so already
+// and places the message in the retry-queue.
+func (t *OutputConfig) failedDelivery(ctx context.Context, event logevent.LogEvent) {
+	// see if we need to send a pause signal
+	if atomic.CompareAndSwapUint32(&t.isInPause, statusDelivering, statusPaused) {
+		goglog.Logger.Debug("outputhttp requesting pause")
+		err := t.control.RequestPause(ctx)
+		if err != nil {
+			goglog.Logger.Error("outputhttp ", err.Error())
+		}
+	}
+	// queue the event for later retry
+	t.queue <- event
+}
+
+// backgroundtask is running in the background and adds new events to the queue, tries to send them out on the RetryInterval.
+func (t *OutputConfig) backgroundtask() {
+	dur := time.Duration(t.RetryInterval) * time.Second
+	ticker := time.NewTicker(dur)
+	defer ticker.Stop()
+	for {
+		select {
+		case event := <-t.queue:
+			if (t.MaxQueueSize > 0 && t.retryqueue.Len() < t.MaxQueueSize) || t.MaxQueueSize == -1 {
+				t.retryqueue.PushBack(event)
+			}
+		case <-ticker.C:
+			// We have reached a RetryInterval. If there are any events in the queue, lets send one back.
+			// If we are still in pause mode we will send one, if we are in normal mode we will send all back.
+			if e := t.retryqueue.Front(); e != nil {
+				if atomic.LoadUint32(&t.isInPause) == statusPaused {
+					event := e.Value.(logevent.LogEvent)
+					t.retryqueue.Remove(e)
+					go func() {
+						ctx, cancel := context.WithTimeout(context.Background(), dur)
+						err := t.OutputEvent(ctx, event)
+						if err != nil {
+							goglog.Logger.Error("outputhttp background sendone ", err.Error())
+						}
+						cancel()
+					}()
```

Comment:
Why using another goroutine to retry?

Reply:
When I send something back to OutputEvent the call can block for a long time in case there are timeout issues present -  this will pause the entire background task until OutputEvent has returned. A goroutine is used to mitigate this.

As already mentioned this is critical code because I need to be able to process new events coming into the queue.